concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -31,4 +28,87 @@ class MonadMulticast m where
 newtype BakerContext = BakerContext
     { _bakerIdentity :: BakerIdentity
     }
-makeClassy ''BakerContext
\ No newline at end of file
+
+makeClassy ''BakerContext
+
+-- |A Monad for timer related actions.
+class MonadTimeout m where
+    -- |Reset the timeout from the supplied 'Duration'.
+    resetTimer :: Duration -> m ()
+
+-- |Return 'Just FinalizerInfo' if the consensus running
+-- is part of the of the provided 'BakersAndFinalizers'.
+-- Otherwise return 'Nothing'.
+isBakerFinalizer ::
+    BakerId ->
+    -- |A collection of bakers and finalizers.
+    BakersAndFinalizers ->
+    -- |'True' if the consensus is part of the finalization committee.
+    -- Otherwise 'False'
+    Maybe FinalizerInfo
+isBakerFinalizer bakerId bakersAndFinalizers = do
+    -- This is O(n) but in principle we could do binary search here as the 'committeeFinalizers' are
+    -- sorted by ascending baker id.
+    Vector.find (\finalizerInfo -> finalizerBakerId finalizerInfo == bakerId) finalizers
+  where
+    finalizers = committeeFinalizers $ bakersAndFinalizers ^. bfFinalizers
+
+-- |Produce a block and multicast it onto the network.
+makeBlock :: MonadState (SkovData (MPV m)) m => m ()
+makeBlock = return ()
+
+-- |Make a block if the consensus runner is leader for the
+-- current round.
+-- TODO: call 'makeBlock' if we're leader for the current round.
+makeBlockIfLeader :: MonadState (SkovData (MPV m)) m => m ()
+makeBlockIfLeader = return ()
+
+-- |Advance to the provided 'Round'.
+--
+-- This function does the following:
+-- * Update the current 'RoundStatus'.
+-- * Persist the new 'RoundStatus'.
+-- * If the consensus runner is leader in the new
+--   round then make the new block.
+advanceRound ::
+    ( MonadReader r m,
+      HasBakerContext r,
+      MonadTimeout m,
+      LowLevel.MonadTreeStateStore m,
+      MonadState (SkovData (MPV m)) m
+    ) =>
+    -- |The 'Round' to progress to.
+    Round ->
+    -- |If we are advancing from a round that timed out
+    -- then this will be @Just 'TimeoutCertificate, 'QuorumCertificate')@ otherwise
+    -- 'Nothing'.
+    --
+    -- In case of the former then the 'TimeoutCertificate' is from the round we're
+    -- advancing from and the associated 'QuorumCertificate' verifies it.
+    Maybe (TimeoutCertificate, QuorumCertificate) ->
+    m ()
+advanceRound newRound timedOut = do
+    myBakerId <- bakerId <$> view bakerIdentity
+    currentRoundStatus <- use roundStatus
+    -- Reset the timeout timer if the consensus runner is part of the
+    -- finalization committee.
+    resetTimerIfFinalizer myBakerId (rsCurrentTimeout currentRoundStatus)
+    -- Advance the round.
+    roundStatus .=! advanceRoundStatus newRound timedOut currentRoundStatus
+    -- Write the new round status to disk.
+    writeCurrentRoundStatus =<< use roundStatus
```

Comment:
We should have a function `setRoundStatus` (in Implementation.hs) that does both things because they should always be kept in sync.

Reply:
I added a setRoundStatus in the Implementation.hs module now.